src/contextmenu.js

```diff
@@ -0,0 +1,472 @@
+      });
+      Blockly.Events.setGroup(false);
+    },
+    scopeType: Blockly.ContextMenuRegistry.ScopeType.BLOCK,
+    id: 'blockDisable',
+    weight: 5,
+  };
+  Blockly.ContextMenuRegistry.registry.register(disableOption);
+};
+
+/**
+ * Modification for context menu 'Delete' to be available for multiple blocks.
+ */
+const registerDelete = function() {
+  const deleteOption = {
+    displayText: function() {
+      let descendantCount = 0;
+      blockSelection.forEach(function(id) {
+        const block = Blockly.getMainWorkspace().getBlockById(id);
+        if (block) {
+          // Count the number of blocks that are nested in this block.
+          descendantCount += block.getDescendants(false).length;
+          const nextBlock = block.getNextBlock();
+          if (nextBlock) {
+            // Blocks in the current stack would survive this block's deletion.
+            descendantCount -= nextBlock.getDescendants(false).length;
+          }
+        }
+      });
+      return (descendantCount === 1) ?
+        Blockly.Msg['DELETE_BLOCK'] :
+        Blockly.Msg['DELETE_X_BLOCKS'].replace('%1', String(descendantCount));
+    },
+    preconditionFn: function(scope) {
+      if (!scope.block.isInFlyout && scope.block.isDeletable()) {
+        return 'enabled';
+      }
+      return 'hidden';
+    },
+    callback: function(scope) {
+      const apply = function(block) {
+        if (block && deleteOption.preconditionFn({
+          block: block,
+        }) === 'enabled' && !hasSelectedParent(block)) {
+          if (block.workspace.isFlyout) {
+            return;
+          }
+          block.workspace.hideChaff();
+          if (block.outputConnection) {
+            block.dispose(false, true);
+          } else {
+            block.dispose(true, true);
+          }
+        }
+      };
+      Blockly.Events.setGroup(true);
+      if (scope.block && blockSelection.size === 0) {
+        apply(scope.block);
+      }
+      blockSelection.forEach(function(id) {
+        const block = Blockly.getMainWorkspace().getBlockById(id);
+        if (block) {
+          apply(block);
+        }
+      });
+      Blockly.Events.setGroup(false);
+    },
+    scopeType: Blockly.ContextMenuRegistry.ScopeType.BLOCK,
+    id: 'blockDelete',
+    weight: 6,
+  };
+  Blockly.ContextMenuRegistry.registry.register(deleteOption);
+};
+
+/**
+ * Add context menu 'Select all Blocks' for workspace.
+ */
+const registerSelectAll = function() {
+  const selectAllOption = {
+    displayText: function() {
+      return 'Select all Blocks';
+    },
+    preconditionFn: function(scope) {
+      const topBlocks = scope.workspace.getTopBlocks(false);
+      for (let i = 0; i < topBlocks.length; i++) {
+        let block = topBlocks[i];
+        while (block) {
+          if (block.isDeletable() &&
+          block.isMovable()) {
+            return 'enabled';
+          }
+          block = block.getNextBlock();
+        }
+      }
+      return 'disabled';
```

Comment:
I think this can be refactored to use `workspace.getAllBlocks()` and [some()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some).

Something like:
```
return scope.workspace.getAllBlocks().some((b) => b.isDeletable() && b.isMovable()) ?
    'enabled' : 'disabled';
```

Reply:
Actually, this precondition does not match your callback. In your callback you only ever select top blocks, even if they aren't movable/deletable and their children are. So I think you want to use `getTopBlocks` in your precondition, and not consider next blocks.

Either that, or modify your callback to match.

Reply:
OKay, so I guess using getTopBlocks would be the best suit. We don't want to select the children if the parents aren't movable/deletable, so that no block connections may get accidentally disrupted.